nemo/collections/asr/modules/dual_path_model.py

```diff
@@ -0,0 +1,321 @@
+        self.end_conv1x1 = nn.Conv1d(feat_out, feat_in, 1, bias=False)
+        self.prelu = nn.PReLU()
+        self.activation = nn.ReLU()
+
+        # gated output layeddr
+        self.output = nn.Sequential(nn.Conv1d(feat_out, feat_out, 1), nn.Tanh())
+        self.output_gate = nn.Sequential(nn.Conv1d(feat_out, feat_out, 1), nn.Sigmoid())
+
+    def forward(self, x):
+        """
+        Return output tensor
+
+        Args:
+            x : torch.Tensor
+                [B, F, N]
+                B: batch
+                F: feat size
+                N: seq length / num of time points
+
+        Returns:
+            out: torch.Tensor
+                [spks, B, F, N]
+        """
+        # norm+ linear
+        x = self.norm(x)
+        x = self.conv1d(x)
+
+        # chunk
+        hop = self.chunk_len // 2
+        x, pad_len = self._chunk(x, self.chunk_len, hop)
+
+        for i, layer in enumerate(self.layers):
+            x = layer(x)
+        x = self.prelu(x)
+        # [B, F, C, Nc]
+
+        x = self.conv2d(x)
+        # [B, F*num_speakers, C, Nc]
+        B, _, C, Nc = x.shape
+
+        x = x.view(B * self.num_speakers, -1, C, Nc)
+        # [B*num_speakers, F, C, Nc]
+
+        x = self._overlap_add(x, pad_len, hop)
+        x = self.output(x) * self.output_gate(x)
+
+        x = self.end_conv1x1(x)
+        # [B*num_speakers, F, N]
+
+        _, F, N = x.shape
+        x = x.view(B, self.num_speakers, F, N)
+        x = self.activation(x)
+
+        x = x.transpose(0, 1)
+        # [num_speakers, B, F, N]
+
+        return x
+
+    def _overlap_add(self, x, pad_len, hop):
+        """
+        Merge through overlap and add
+        """
+
+        B, F, C, Nc = x.shape
+
+        x = x.transpose(2, 3).contiguous().view(B, F, -1, C * 2)
+
+        x1 = x[:, :, :, :C].contiguous().view(B, F, -1)[:, :, hop:]
+        x2 = x[:, :, :, C:].contiguous().view(B, F, -1)[:, :, :-hop]
+        x = x1 + x2
+
+        if pad_len > 0:
+            x = x[:, :, :-pad_len]
+
+        return x
+
+    def _chunk(self, x, chunk_len, hop):
+        """
+        Segment and stack encoder output
+        
+        Args:
+            x: torch.Tensor
+                [B, F, N]
+                [B, N ,L]
+            chunk_len: length of chunks
+            hop : hop size
+
+        Return:
+            output: torch.Tensor
+                [B, F, C, Nc]
+            pad: padding used 
+        """
+        B, F, N = x.shape
+        x, pad_len = self._padding(x, chunk_len, hop)
+        x1 = x[:, :, :-hop].contiguous().view(B, F, -1, chunk_len)
+        x2 = x[:, :, hop:].contiguous().view(B, F, -1, chunk_len)
+        x = torch.cat([x1, x2], dim=3).view(B, F, -1, chunk_len).transpose(2, 3)
+        return x.contiguous(), pad_len
+
+    def _padding(self, x, chunk_len, hop):
+        """
+        pad for whole number of chunks
+        """
+        B, F, N = x.shape
+        pad_len = chunk_len - (hop + N % chunk_len) % chunk_len
```

Comment:
I'm not sure if i'm mistaken. 

But if c=4, h= 2, N=7, then pad_len should be 1 right? but this computes 3 for some reason